File: database-driver-plugin/src/main/java/de/lino/database/utility/cache/provider/DefaultCacheProvider.java
package de.lino.database.utility.cache.provider;

import de.lino.database.utility.cache.DefaultCache;
import de.lino.database.utility.cache.DefaultClusteredCache;
import de.lino.database.utils.cache.Cache;
import de.lino.database.utils.cache.provider.CacheProvider;
import de.lino.database.utils.cache.provider.Caches;
import de.lino.database.utils.cache.ClusteredCache;

import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

/**
 * Default {@link CacheProvider}, backing {@link Caches} with {@link DefaultCache}
 * and {@link DefaultClusteredCache}. Registered as a service in
 * {@code META-INF/services/de.lino.database.utils.cache.provider.CacheProvider} so it is picked
 * up automatically by {@link java.util.ServiceLoader} once this module is on the classpath.
 */
public final class DefaultCacheProvider implements CacheProvider {

    @Override
    public <ID, T> Cache<ID, T> newCache(Function<ID, CompletableFuture<T>> loader, Duration ttl, long maxSize) {
        return new DefaultCache<>(loader, ttl, maxSize);
    }

    @Override
    public <ID, T> ClusteredCache<ID, T> newClusteredCache(int shardCount, int replicationFactor,
                                                           Function<ID, CompletableFuture<T>> loader,
                                                           Duration ttl, long maxSizePerShard) {
        return new DefaultClusteredCache<>(shardCount, replicationFactor, loader, ttl, maxSizePerShard);
    }
}

File: python/database-driver-plugin/src/database_driver/plugin/utility/cache/provider/default_cache_provider.py
"""Mirror of ``de.lino.database.utility.cache.provider.DefaultCacheProvider``."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from datetime import timedelta
from typing import Any

from database_driver.api.utils.cache.cache import Cache
from database_driver.api.utils.cache.clustered_cache import ClusteredCache
from database_driver.api.utils.cache.provider.cache_provider import CacheProvider

from database_driver.plugin.utility.cache.default_cache import DefaultCache
from database_driver.plugin.utility.cache.default_clustered_cache import DefaultClusteredCache


class DefaultCacheProvider(CacheProvider):
    """Default ``CacheProvider``, backing the api package's ``Caches`` with
    ``DefaultCache`` and ``DefaultClusteredCache``. Registered under the
    ``database_driver.cache_provider`` entry-point group in this distribution's
    packaging metadata - the Python analogue of the Java module's
    ``META-INF/services`` resource - so it is picked up automatically once this
    distribution is installed."""

    def new_cache(
        self,
        loader: Callable[[Any], Awaitable[Any]],
        ttl: timedelta | None,
        max_size: int,
    ) -> Cache[Any, Any]:
        return DefaultCache(loader, ttl, max_size)

    def new_clustered_cache(
        self,
        shard_count: int,
        replication_factor: int,
        loader: Callable[[Any], Awaitable[Any]],
        ttl: timedelta | None,
        max_size_per_shard: int,
    ) -> ClusteredCache[Any, Any]:
        return DefaultClusteredCache(shard_count, replication_factor, loader, ttl, max_size_per_shard)
